File: user-service/src/test/java/com/github/manosbatsis/userservice/mapper/UserMapperTest.java
package com.github.manosbatsis.userservice.mapper;

import com.github.javafaker.Faker;
import com.github.manosbatsis.userservice.mapper.UserMapper;
import com.github.manosbatsis.userservice.model.User;
import com.github.manosbatsis.userservice.rest.dto.CreateUserRequest;
import com.github.manosbatsis.userservice.rest.dto.UpdateUserRequest;
import com.github.manosbatsis.userservice.rest.dto.UserResponse;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Import;
import org.springframework.test.context.junit.jupiter.SpringExtension;

import java.util.stream.Stream;

import static org.assertj.core.api.Assertions.assertThat;

@ExtendWith(SpringExtension.class)
@Import(UserMapperImpl.class)
class UserMapperTest {

    @Autowired
    private UserMapper userMapper;

    private Faker faker = new Faker();

    @Test
    void testToUser() {
        CreateUserRequest createUserRequest = new CreateUserRequest(
            faker.internet().emailAddress(),
            faker.name().fullName(),
            faker.address().fullAddress(),
            true
        );

        User user = userMapper.toUser(createUserRequest);

        assertThat(user.getId()).isNull();
        assertThat(user.getEmail()).isEqualTo(createUserRequest.getEmail());
        assertThat(user.getFullName()).isEqualTo(createUserRequest.getFullName());
        assertThat(user.getActive()).isEqualTo(createUserRequest.getActive());
        assertThat(user.getCreatedAt()).isNull();
        assertThat(user.getUpdatedAt()).isNull();
    }

    @Test
    void testToUserResponse() {
        User user = new User(
            faker.internet().emailAddress(),
            faker.name().fullName(),
            faker.address().fullAddress(),
            true
        );

        UserResponse userResponse = userMapper.toUserResponse(user);

        assertThat(userResponse.id()).isNull();
        assertThat(userResponse.email()).isEqualTo(user.getEmail());
        assertThat(userResponse.fullName()).isEqualTo(user.getFullName());
        assertThat(userResponse.active()).isEqualTo(user.getActive());
        assertThat(user.getCreatedAt()).isNull();
        assertThat(user.getUpdatedAt()).isNull();
    }

    @ParameterizedTest
    @MethodSource("provideUpdateUserFromRequest")
    void testUpdateUserFromRequest(String newFullName, Boolean newActive, User expectedUser) {

        User user = new User("email@test", "fullName", "address", true);

        UpdateUserRequest updateUserRequest = UpdateUserRequest.builder()
            .fullName(newFullName)
            .active(newActive)
            .build();

        userMapper.updateUserFromRequest(updateUserRequest, user);
        assertThat(user).isEqualTo(expectedUser);
    }

    private static Stream<Arguments> provideUpdateUserFromRequest() {

        return Stream.of(
                Arguments.of("fullName2", false, new User("email@test", "fullName2", "address", false)),
                Arguments.of(null, null, new User("email@test", "fullName", "address", true)),
                Arguments.of("fullName2", null, new User("email@test", "fullName2", "address", true)),
                Arguments.of(null, false, new User("email@test", "fullName", "address", false)),
                Arguments.of(null, null, new User("email@test", "fullName", "address", true))
        );
    }
}

File: user-service/src/test/kotlin/com/github/manosbatsis/services/user/mapper/UserMapperTest.kt
package com.github.manosbatsis.services.user.mapper

import com.github.javafaker.Faker
import com.github.manosbatsis.services.user.model.User
import com.github.manosbatsis.services.user.rest.dto.CreateUserRequest
import com.github.manosbatsis.services.user.rest.dto.UpdateUserRequest
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.context.junit.jupiter.SpringExtension
import java.util.stream.Stream

@ExtendWith(SpringExtension::class)
@ActiveProfiles(profiles = ["test"])
class UserMapperTest {
    private val userMapper = UserMapper.instance()
    private val faker = Faker()

    @Test
    fun testToUser() {
        val createUserRequest = CreateUserRequest(
            faker.internet().emailAddress(),
            faker.name().fullName(),
            faker.address().fullAddress(),
            true,
        )
        val user = userMapper.toUser(createUserRequest)
        assertThat(user.id).isEqualTo(0L)
        assertThat(user.email).isEqualTo(createUserRequest.email)
        assertThat(user.fullName).isEqualTo(createUserRequest.fullName)
        assertThat(user.active).isEqualTo(createUserRequest.active)
        assertThat(user.createdAt).isNull()
        assertThat(user.updatedAt).isNull()
    }

    @Test
    fun testToUserResponse() {
        val user = User(
            email = faker.internet().emailAddress(),
            fullName = faker.name().fullName(),
            address = faker.address().fullAddress(),
            active = true,
        )
        val (id, email, fullName, _, active) = userMapper.toUserResponse(user)
        assertThat(id).isEqualTo(0L)
        assertThat(email).isEqualTo(user.email)
        assertThat(fullName).isEqualTo(user.fullName)
        assertThat(active).isEqualTo(user.active)
        assertThat(user.createdAt).isNull()
        assertThat(user.updatedAt).isNull()
    }

    @ParameterizedTest
    @MethodSource("provideUpdateUserFromRequest")
    fun testUpdateUserFromRequest(newFullName: String?, newActive: Boolean?, expectedUser: User) {
        val user = User(
            email = "email@test",
            fullName = "fullName",
            address = "address",
            active = true,
        )
        val updateUserRequest = UpdateUserRequest(
            fullName = newFullName,
            active = newActive,
        )
        userMapper.updateUserFromRequest(updateUserRequest, user)
        assertThat(user.email).isEqualTo(expectedUser.email)
    }

    companion object {
        @JvmStatic
        private fun provideUpdateUserFromRequest(): Stream<Arguments> {
            return Stream.of(
                Arguments.of(
                    "fullName2",
                    false,
                    User(email = "email@test", fullName = "fullName2", address = "address", active = false),
                ),
                Arguments.of(null, null, User(email = "email@test", fullName = "fullName", address = "address", active = true)),
                Arguments.of(
                    "fullName2",
                    null,
                    User(email = "email@test", fullName = "fullName2", address = "address", active = true),
                ),
                Arguments.of(null, false, User(email = "email@test", fullName = "fullName", address = "address", active = false)),
                Arguments.of(
                    null,
                    null,
                    User(
                        email = "email@test",
                        fullName = "fullName",
                        address = "address",
                        active = true,
                    ),
                ),
            )
        }
    }
}
